// remotion-bits/dist/components/Scene3D/Element3D.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { random } from "remotion";
import { useCamera } from "./context.js";
import { useMotionTiming, buildMotionStyles, getEasingFunction, interpolateKeyframes, } from "../../utils/motion/index.js";
import { Transform3D, Vector3 } from "../../utils/transform3d.js";
import { transformToCSS } from "../../utils/interpolate3d.js";
const ELEMENT3D_SYMBOL = Symbol("Scene3D.Element3D");
export function isElement3D(element) {
    return (React.isValidElement(element) &&
        typeof element.type === "function" &&
        ELEMENT3D_SYMBOL in element.type);
}
function calculateStaggerIndex(actualIndex, totalChildren, direction) {
    if (direction === "reverse") {
        return totalChildren - 1 - actualIndex;
    }
    if (direction === "center") {
        const mid = Math.floor(totalChildren / 2);
        return Math.abs(actualIndex - mid);
    }
    if (direction === "random") {
        const indices = Array.from({ length: totalChildren }, (_, i) => i);
        for (let i = indices.length - 1; i > 0; i--) {
            const j = Math.floor(random(`element3d-stagger-${i}`) * (i + 1));
            [indices[i], indices[j]] = [indices[j], indices[i]];
        }
        return indices.indexOf(actualIndex);
    }
    return actualIndex;
}
function useTransitionStyle(transition) {
    const progress = useMotionTiming({
        frames: transition?.frames,
        duration: transition?.duration,
        delay: transition?.delay ?? 0,
        easing: transition?.easing,
    });
    if (!transition) {
        return {};
    }
    const duration = transition.frames
        ? transition.frames[1] - transition.frames[0]
        : (transition.duration ?? 30);
    return buildMotionStyles({
        progress,
        transforms: {
            x: transition.x,
            y: transition.y,
            z: transition.z,
            scale: transition.scale,
            scaleX: transition.scaleX,
            scaleY: transition.scaleY,
            rotate: transition.rotate,
            rotateX: transition.rotateX,
            rotateY: transition.rotateY,
            rotateZ: transition.rotateZ,
            skew: transition.skew,
            skewX: transition.skewX,
            skewY: transition.skewY,
            transform: transition.transform,
        },
        styles: {
            opacity: transition.opacity,
            color: transition.color,
            backgroundColor: transition.backgroundColor,
            blur: transition.blur,
        },
        easing: getEasingFunction(transition.easing),
        duration,
    });
}
const StaggeredChild = ({ child, actualIndex, staggerIndex, transition, }) => {
    const duration = transition.frames
        ? transition.frames[1] - transition.frames[0]
        : (transition.duration ?? 30);
    const staggerProgress = useMotionTiming({
        frames: transition.frames,
        duration: transition.duration,
        delay: transition.delay ?? 0,
        stagger: transition.stagger,
        unitIndex: staggerIndex,
        easing: transition.easing,
    });
    const staggerStyle = buildMotionStyles({
        progress: staggerProgress,
        transforms: {
            x: transition.x,
            y: transition.y,
            z: transition.z,
            scale: transition.scale,
            scaleX: transition.scaleX,
            scaleY: transition.scaleY,
            rotate: transition.rotate,
            rotateX: transition.rotateX,
            rotateY: transition.rotateY,
            rotateZ: transition.rotateZ,
            skew: transition.skew,
            skewX: transition.skewX,
            skewY: transition.skewY,
            transform: transition.transform,
        },
        styles: {
            opacity: transition.opacity,
            color: transition.color,
            backgroundColor: transition.backgroundColor,
            blur: transition.blur,
        },
        easing: getEasingFunction(transition.easing),
        duration,
    });
    if (React.isValidElement(child)) {
        const existingStyle = child.props.style || {};
        return React.cloneElement(child, {
            key: actualIndex,
            style: { ...existingStyle, ...staggerStyle },
        });
    }
    return (_jsx("span", { style: staggerStyle, children: child }, actualIndex));
};
const Element3DComponent = ({ x = 0, y = 0, z = 0, scale = 1, scaleX = 1, scaleY = 1, rotateX = 0, rotateY = 0, rotateZ = 0, rotateOrder = "xyz", fixed = false, centered = false, transition, className, style, children, }) => {
    const camera = useCamera();
    const xVal = interpolateKeyframes(x, 1);
    const yVal = interpolateKeyframes(y, 1);
    const zVal = interpolateKeyframes(z, 1);
    const scaleVal = interpolateKeyframes(scale, 1);
    const scaleXVal = interpolateKeyframes(scaleX, 1);
    const scaleYVal = interpolateKeyframes(scaleY, 1);
    const rotateXVal = interpolateKeyframes(rotateX, 1);
    const rotateYVal = interpolateKeyframes(rotateY, 1);
    const rotateZVal = interpolateKeyframes(rotateZ, 1);
    const degreesToRadians = Math.PI / 180;
    const elementTransform = Transform3D.fromEuler(rotateXVal * degreesToRadians, rotateYVal * degreesToRadians, rotateZVal * degreesToRadians, new Vector3(xVal, yVal, zVal), new Vector3(scaleVal * scaleXVal, scaleVal * scaleYVal, scaleVal), rotateOrder.toUpperCase());
    let finalTransform;
    if (fixed) {
        const cameraTransform = Transform3D.fromEuler(camera.rotateX * degreesToRadians, camera.rotateY * degreesToRadians, camera.rotateZ * degreesToRadians, new Vector3(camera.x, camera.y, camera.z), new Vector3(camera.scale * camera.scaleX, camera.scale * camera.scaleY, camera.scale));
        finalTransform = cameraTransform.multiply(elementTransform);
    }
    else {
        finalTransform = elementTransform;
    }
    let transformString = transformToCSS(finalTransform);
    if (centered) {
        const centerTransform = "translate(-50%, -50%) rotate(0.01deg)";
        transformString = `${centerTransform} ${transformString}`;
    }
    const childArray = React.Children.toArray(children);
    const totalChildren = childArray.length;
    const transitionStyle = useTransitionStyle(transition?.stagger ? undefined : transition);
    const elementStyle = {
        position: "absolute",
        transformStyle: "preserve-3d",
        transform: transformString,
        ...style,
        ...transitionStyle,
    };
    const renderChildren = () => {
        if (!transition?.stagger) {
            return children;
        }
        const staggerDirection = transition.staggerDirection ?? "forward";
        return childArray.map((child, actualIndex) => {
            const staggerIndex = calculateStaggerIndex(actualIndex, totalChildren, staggerDirection);
            return (_jsx(StaggeredChild, { child: child, actualIndex: actualIndex, staggerIndex: staggerIndex, transition: transition }, actualIndex));
        });
    };
    return (_jsx("div", { className: className, style: elementStyle, "data-element3d-fixed": fixed, children: renderChildren() }));
};
Element3DComponent[ELEMENT3D_SYMBOL] = true;
export const Element3D = Element3DComponent;
