// remotion-bits/dist/components/Scene3D/Scene3D.js
import { jsxs as _jsxs, jsx as _jsx } from "react/jsx-runtime";
import React, { useMemo, useCallback, useRef } from "react";
import { useCurrentFrame, useVideoConfig } from "remotion";
import { Scene3DContext } from "./context.js";
import { isStepElement } from "./Step.js";
import { getEasingFunction, interpolateKeyframes } from "../../utils/motion/index.js";
import { Transform3D, Vector3 } from "../../utils/transform3d.js";
import { interpolateTransform, transformToCSS } from "../../utils/interpolate3d.js";
function stepConfigToForwardTransform(step) {
    const x = interpolateKeyframes(step.x ?? 0, 1);
    const y = interpolateKeyframes(step.y ?? 0, 1);
    const z = interpolateKeyframes(step.z ?? 0, 1);
    const rotateX = interpolateKeyframes(step.rotateX ?? 0, 1);
    const rotateY = interpolateKeyframes(step.rotateY ?? 0, 1);
    const rotateZ = interpolateKeyframes(step.rotateZ ?? 0, 1);
    const scale = interpolateKeyframes(step.scale ?? 1, 1);
    const scaleX = interpolateKeyframes(step.scaleX ?? 1, 1);
    const scaleY = interpolateKeyframes(step.scaleY ?? 1, 1);
    const scaleZ = interpolateKeyframes(step.scaleZ ?? 1, 1);
    const degreesToRadians = Math.PI / 180;
    return Transform3D.fromEuler(rotateX * degreesToRadians, rotateY * degreesToRadians, rotateZ * degreesToRadians, new Vector3(x, y, z), new Vector3(scale * scaleX, scale * scaleY, scale * scaleZ), step.rotateOrder?.toUpperCase() ?? 'XYZ');
}
function stepConfigToCameraTransform(step, fitScale) {
    const forward = stepConfigToForwardTransform(step);
    const inverse = forward.inverse();
    inverse.position.multiplyScalar(fitScale);
    inverse.scale.multiplyScalar(fitScale);
    return inverse;
}
function stepConfigToCameraTarget(step) {
    return {
        x: interpolateKeyframes(step.x ?? 0, 1),
        y: interpolateKeyframes(step.y ?? 0, 1),
        z: interpolateKeyframes(step.z ?? 0, 1),
        rotateX: interpolateKeyframes(step.rotateX ?? 0, 1),
        rotateY: interpolateKeyframes(step.rotateY ?? 0, 1),
        rotateZ: interpolateKeyframes(step.rotateZ ?? 0, 1),
        scale: interpolateKeyframes(step.scale ?? 1, 1),
        scaleX: interpolateKeyframes(step.scaleX ?? 1, 1),
        scaleY: interpolateKeyframes(step.scaleY ?? 1, 1),
        scaleZ: interpolateKeyframes(step.scaleZ ?? 1, 1),
    };
}
export const Scene3D = ({ perspective = 1000, transitionDuration = 30, easing = "easeInOutCubic", activeStep, stepDuration, width, height, className, style, children, }) => {
    const frame = useCurrentFrame();
    const { durationInFrames, width: videoWidth, height: videoHeight } = useVideoConfig();
    const fitScale = useMemo(() => {
        const designWidth = width ?? videoWidth;
        const designHeight = height ?? videoHeight;
        const xScale = videoWidth / designWidth;
        const yScale = videoHeight / designHeight;
        return Math.min(xScale, yScale);
    }, [width, height, videoWidth, videoHeight]);
    const stepsRef = useRef([]);
    const stepIndexRef = useRef(0);
    const childArray = React.Children.toArray(children);
    const stepChildren = useMemo(() => childArray.filter(isStepElement), [childArray]);
    const nonStepChildren = useMemo(() => childArray.filter((child) => !isStepElement(child)), [childArray]);
    const steps = useMemo(() => {
        const stepConfigs = [];
        let currentFrame = 0;
        stepChildren.forEach((child, index) => {
            const props = child.props;
            const duration = props.duration ?? stepDuration ?? Math.floor(durationInFrames / stepChildren.length);
            const enterFrame = currentFrame;
            const exitFrame = currentFrame + duration;
            stepConfigs.push({
                id: props.id ?? `step-${index}`,
                index,
                x: props.x ?? 0,
                y: props.y ?? 0,
                z: props.z ?? 0,
                scale: props.scale ?? 1,
                scaleX: props.scaleX ?? 1,
                scaleZ: props.scaleZ ?? 1,
                scaleY: props.scaleY ?? 1,
                rotateX: props.rotateX ?? 0,
                rotateY: props.rotateY ?? 0,
                rotateZ: props.rotateZ ?? 0,
                rotateOrder: props.rotateOrder ?? "xyz",
                enterFrame,
                exitFrame,
            });
            currentFrame = exitFrame;
        });
        stepsRef.current = stepConfigs;
        return stepConfigs;
    }, [stepChildren, stepDuration, durationInFrames]);
    const registerStep = useCallback((config) => {
        const index = stepIndexRef.current;
        stepIndexRef.current += 1;
        return index;
    }, []);
    const { activeStepIndex, transitionProgress, camera, cameraTransform } = useMemo(() => {
        if (steps.length === 0) {
            return {
                activeStepIndex: 0,
                transitionProgress: 0,
                camera: { x: 0, y: 0, z: 0, rotateX: 0, rotateY: 0, rotateZ: 0, scale: 1, scaleX: 1, scaleY: 1, scaleZ: 1 },
                cameraTransform: Transform3D.identity(),
            };
        }
        let currentStepIndex;
        if (activeStep !== undefined) {
            if (typeof activeStep === "number") {
                currentStepIndex = Math.max(0, Math.min(activeStep, steps.length - 1));
            }
            else {
                const foundIndex = steps.findIndex((s) => s.id === activeStep);
                currentStepIndex = foundIndex >= 0 ? foundIndex : 0;
            }
        }
        else {
            currentStepIndex = steps.findIndex((step) => frame >= step.enterFrame && frame < step.exitFrame);
            if (currentStepIndex < 0) {
                currentStepIndex = frame >= steps[steps.length - 1].exitFrame ? steps.length - 1 : 0;
            }
        }
        const currentStep = steps[currentStepIndex];
        const prevStep = currentStepIndex > 0 ? steps[currentStepIndex - 1] : null;
        const transitionStartFrame = currentStep.enterFrame;
        const transitionEndFrame = transitionStartFrame + transitionDuration;
        let progress;
        if (frame < transitionStartFrame) {
            progress = 0;
        }
        else if (frame >= transitionEndFrame) {
            progress = 1;
        }
        else {
            progress = (frame - transitionStartFrame) / transitionDuration;
        }
        const easingFn = getEasingFunction(easing);
        const easedProgress = easingFn ? easingFn(progress) : progress;
        const targetCamera = stepConfigToCameraTarget(currentStep);
        const targetTransform = stepConfigToCameraTransform(currentStep, fitScale);
        let cameraState;
        let cameraTransformState;
        if (prevStep && progress < 1) {
            const fromCamera = stepConfigToCameraTarget(prevStep);
            const fromForward = stepConfigToForwardTransform(prevStep);
            const targetForward = stepConfigToForwardTransform(currentStep);
            const currentForward = interpolateTransform(fromForward, targetForward, easedProgress, easingFn);
            cameraTransformState = currentForward.inverse();
            cameraTransformState.position.multiplyScalar(fitScale);
            cameraTransformState.scale.multiplyScalar(fitScale);
            cameraState = {
                x: fromCamera.x + (targetCamera.x - fromCamera.x) * easedProgress,
                y: fromCamera.y + (targetCamera.y - fromCamera.y) * easedProgress,
                z: fromCamera.z + (targetCamera.z - fromCamera.z) * easedProgress,
                rotateX: fromCamera.rotateX + (targetCamera.rotateX - fromCamera.rotateX) * easedProgress,
                rotateY: fromCamera.rotateY + (targetCamera.rotateY - fromCamera.rotateY) * easedProgress,
                rotateZ: fromCamera.rotateZ + (targetCamera.rotateZ - fromCamera.rotateZ) * easedProgress,
                scale: fromCamera.scale + (targetCamera.scale - fromCamera.scale) * easedProgress,
                scaleX: fromCamera.scaleX + (targetCamera.scaleX - fromCamera.scaleX) * easedProgress,
                scaleY: fromCamera.scaleY + (targetCamera.scaleY - fromCamera.scaleY) * easedProgress,
                scaleZ: fromCamera.scaleZ + (targetCamera.scaleZ - fromCamera.scaleZ) * easedProgress,
            };
        }
        else {
            cameraState = targetCamera;
            cameraTransformState = targetTransform;
        }
        return {
            activeStepIndex: currentStepIndex,
            transitionProgress: easedProgress,
            camera: cameraState,
            cameraTransform: cameraTransformState,
        };
    }, [steps, frame, activeStep, transitionDuration, easing, fitScale]);
    const activeStepId = steps[activeStepIndex]?.id;
    const contextValue = useMemo(() => ({
        camera,
        activeStepIndex,
        activeStepId,
        transitionProgress,
        transitionDuration,
        steps,
        registerStep,
    }), [camera, activeStepIndex, activeStepId, transitionProgress, transitionDuration, steps, registerStep]);
    const containerStyle = {
        position: "relative",
        width: "100%",
        height: "100%",
        overflow: "hidden",
        perspective: `${perspective}px`,
        ...style,
    };
    const canvasStyle = {
        position: "absolute",
        width: "100%",
        height: "100%",
        transformStyle: "preserve-3d",
        transform: transformToCSS(cameraTransform),
        transformOrigin: "center center",
    };
    const worldOriginStyle = {
        position: "absolute",
        left: "50%",
        top: "50%",
        transformStyle: "preserve-3d",
    };
    return (_jsx(Scene3DContext.Provider, { value: contextValue, children: _jsx("div", { className: className, style: containerStyle, "data-scene3d": true, children: _jsx("div", { style: canvasStyle, "data-scene3d-canvas": true, children: _jsxs("div", { style: worldOriginStyle, "data-scene3d-world": true, children: [stepChildren, nonStepChildren] }) }) }) }));
};
