// remotion-bits/dist/components/Scene3D/Step.js
import { jsx as _jsx } from "react/jsx-runtime";
import React, { useLayoutEffect, useMemo } from "react";
import { random } from "remotion";
import { useScene3D } from "./context.js";
import { useMotionTiming, buildMotionStyles, getEasingFunction, interpolateKeyframes, StepTimingContext, } from "../../utils/motion/index.js";
import { Transform3D, Vector3 } from "../../utils/transform3d.js";
import { transformToCSS } from "../../utils/interpolate3d.js";
const STEP_SYMBOL = Symbol("Scene3D.Step");
export function isStepElement(element) {
    return (React.isValidElement(element) &&
        typeof element.type === "function" &&
        STEP_SYMBOL in element.type);
}
function calculateStaggerIndex(actualIndex, totalChildren, direction) {
    if (direction === "reverse") {
        return totalChildren - 1 - actualIndex;
    }
    if (direction === "center") {
        const mid = Math.floor(totalChildren / 2);
        return Math.abs(actualIndex - mid);
    }
    if (direction === "random") {
        const indices = Array.from({ length: totalChildren }, (_, i) => i);
        for (let i = indices.length - 1; i > 0; i--) {
            const j = Math.floor(random(`step-stagger-${i}`) * (i + 1));
            [indices[i], indices[j]] = [indices[j], indices[i]];
        }
        return indices.indexOf(actualIndex);
    }
    return actualIndex;
}
function getTransitionStyles(transition, progress, duration) {
    return buildMotionStyles({
        progress,
        transforms: {
            x: transition.x,
            y: transition.y,
            z: transition.z,
            scale: transition.scale,
            scaleX: transition.scaleX,
            scaleY: transition.scaleY,
            rotate: transition.rotate,
            rotateX: transition.rotateX,
            rotateY: transition.rotateY,
            rotateZ: transition.rotateZ,
            skew: transition.skew,
            skewX: transition.skewX,
            skewY: transition.skewY,
            transform: transition.transform,
        },
        styles: {
            opacity: transition.opacity,
            color: transition.color,
            backgroundColor: transition.backgroundColor,
            blur: transition.blur,
        },
        easing: getEasingFunction(transition.easing),
        duration,
    });
}
function useTransitionStyle(transition, shouldApply) {
    const duration = transition?.frames
        ? transition.frames[1] - transition.frames[0]
        : (transition?.duration ?? 30); // Default based on fallback logic logic
    const progress = useMotionTiming({
        frames: transition?.frames,
        duration: transition?.duration,
        delay: transition?.delay ?? 0,
        easing: transition?.easing,
    });
    if (!transition || !shouldApply) {
        return {};
    }
    return getTransitionStyles(transition, progress, duration);
}
const StaggeredChild = ({ child, actualIndex, staggerIndex, transition, }) => {
    const duration = transition.frames
        ? transition.frames[1] - transition.frames[0]
        : (transition.duration ?? 30);
    const staggerProgress = useMotionTiming({
        frames: transition.frames,
        duration: transition.duration,
        delay: transition.delay ?? 0,
        stagger: transition.stagger,
        unitIndex: staggerIndex,
        easing: transition.easing,
    });
    const staggerStyle = buildMotionStyles({
        progress: staggerProgress,
        transforms: {
            x: transition.x,
            y: transition.y,
            z: transition.z,
            scale: transition.scale,
            scaleX: transition.scaleX,
            scaleY: transition.scaleY,
            rotate: transition.rotate,
            rotateX: transition.rotateX,
            rotateY: transition.rotateY,
            rotateZ: transition.rotateZ,
            skew: transition.skew,
            skewX: transition.skewX,
            skewY: transition.skewY,
            transform: transition.transform,
        },
        styles: {
            opacity: transition.opacity,
            color: transition.color,
            backgroundColor: transition.backgroundColor,
            blur: transition.blur,
        },
        easing: getEasingFunction(transition.easing),
        duration,
    });
    if (React.isValidElement(child)) {
        const existingStyle = child.props.style || {};
        return React.cloneElement(child, {
            key: actualIndex,
            style: { ...existingStyle, ...staggerStyle },
        });
    }
    return (_jsx("span", { style: staggerStyle, children: child }, actualIndex));
};
const StepComponent = ({ id, duration, x = 0, y = 0, z = 0, scale = 1, scaleX = 1, scaleY = 1, scaleZ = 1, rotateX = 0, rotateY = 0, rotateZ = 0, rotateOrder = "xyz", transition, exitTransition, className, style, children, }) => {
    const { activeStepIndex, steps, registerStep } = useScene3D();
    useLayoutEffect(() => {
        registerStep({
            id,
            x,
            y,
            z,
            scale,
            scaleX,
            scaleY,
            scaleZ,
            rotateX,
            rotateY,
            rotateZ,
            rotateOrder,
            enterFrame: 0,
            exitFrame: 0,
        });
    }, []);
    const stepIndex = steps.findIndex((s) => s.id === id);
    const isActive = stepIndex === activeStepIndex;
    const isPast = stepIndex < activeStepIndex;
    const xVal = interpolateKeyframes(x, 1);
    const yVal = interpolateKeyframes(y, 1);
    const zVal = interpolateKeyframes(z, 1);
    const scaleVal = interpolateKeyframes(scale, 1);
    const scaleXVal = interpolateKeyframes(scaleX, 1);
    const scaleZVal = interpolateKeyframes(scaleZ, 1);
    const scaleYVal = interpolateKeyframes(scaleY, 1);
    const rotateXVal = interpolateKeyframes(rotateX, 1);
    const rotateYVal = interpolateKeyframes(rotateY, 1);
    const rotateZVal = interpolateKeyframes(rotateZ, 1);
    const degreesToRadians = Math.PI / 180;
    const stepTransform = Transform3D.fromEuler(rotateXVal * degreesToRadians, rotateYVal * degreesToRadians, rotateZVal * degreesToRadians, new Vector3(xVal, yVal, zVal), new Vector3(scaleVal * scaleXVal, scaleVal * scaleYVal, scaleVal * scaleZVal), rotateOrder.toUpperCase());
    const centerTransform = Transform3D.identity()
        .translate(-0.5, -0.5, 0)
        .rotateZ(0.01 * degreesToRadians);
    const positionTransform = `translate(-50%, -50%) rotate(0.01deg) ${transformToCSS(stepTransform)}`;
    const stepConfig = steps[stepIndex];
    const computedEnterTransition = useMemo(() => {
        if (!transition)
            return undefined;
        if (transition.frames)
            return transition;
        if (!stepConfig)
            return transition;
        const startFrame = stepConfig.enterFrame;
        const duration = transition.duration ?? 30;
        return {
            ...transition,
            frames: [startFrame, startFrame + duration],
        };
    }, [transition, stepConfig]);
    const computedExitTransition = useMemo(() => {
        if (!exitTransition)
            return undefined;
        if (exitTransition.frames)
            return exitTransition;
        if (!stepConfig)
            return exitTransition;
        const startFrame = stepConfig.exitFrame;
        const duration = exitTransition.duration ?? 30;
        return {
            ...exitTransition,
            frames: [startFrame, startFrame + duration],
        };
    }, [exitTransition, stepConfig]);
    const enterTransitionStyle = useTransitionStyle(computedEnterTransition, !isPast);
    const exitTransitionStyle = useTransitionStyle(computedExitTransition, isPast);
    const activeTransitionStyle = isPast
        ? exitTransitionStyle
        : enterTransitionStyle;
    const childArray = React.Children.toArray(children);
    const totalChildren = childArray.length;
    const stepStyle = {
        position: "absolute",
        transformStyle: "preserve-3d",
        transform: positionTransform,
        ...style,
        ...activeTransitionStyle,
    };
    const renderChildren = () => {
        if (!computedEnterTransition?.stagger) {
            return children;
        }
        const staggerDirection = computedEnterTransition.staggerDirection ?? "forward";
        return childArray.map((child, actualIndex) => {
            const staggerIndex = calculateStaggerIndex(actualIndex, totalChildren, staggerDirection);
            return (_jsx(StaggeredChild, { child: child, actualIndex: actualIndex, staggerIndex: staggerIndex, transition: computedEnterTransition }, actualIndex));
        });
    };
    return (_jsx("div", { className: className, style: stepStyle, "data-step-id": id, "data-step-active": isActive, children: _jsx(StepTimingContext.Provider, { value: { stepConfig }, children: renderChildren() }) }));
};
StepComponent[STEP_SYMBOL] = true;
export const Step = StepComponent;
