// remotion-bits/dist/utils/motion/index.js
import { useCurrentFrame, useVideoConfig } from "remotion";
import { Easing, interpolate } from "../interpolate.js";
import { matrixToCSS } from "../interpolate3d.js";
import { interpolateColorKeyframes } from "../color.js";
import { useStepTiming } from "../StepContext.js";
export function interpolateKeyframes(value, progress, easingFn, duration) {
    if (!Array.isArray(value))
        return value;
    const rawKeyframes = value;
    if (rawKeyframes.length === 0)
        return 0;
    // Check for holds
    const hasHolds = rawKeyframes.some(k => typeof k === 'object' && k && k.type === 'hold');
    if (hasHolds) {
        if (typeof duration !== 'number') {
            console.warn("interpolateKeyframes: duration required for hold frames");
            return rawKeyframes.filter(k => !(typeof k === 'object' && k && k.type === 'hold'))[0];
        }
        const simpleKeyframes = [];
        const inputRange = [];
        // Calculate timing
        const holds = rawKeyframes.filter(k => typeof k === 'object' && k && k.type === 'hold');
        const totalHoldFrames = holds.reduce((acc, h) => acc + h.frames, 0);
        // Calculate number of transitions
        // A transition occurs between two values.
        // Holds are inserted "at" a value (extending it).
        // Sequence: Value1, Hold1, Value2, Hold2, Value3...
        // Transitions: V1->V2, V2->V3.
        // Total duration available for transitions = duration - totalHoldFrames
        const valuesOnly = rawKeyframes.filter(k => !(typeof k === 'object' && k && k.type === 'hold'));
        const transitionsCount = Math.max(0, valuesOnly.length - 1);
        const availableFrames = Math.max(0, duration - totalHoldFrames);
        const transitionDuration = transitionsCount > 0 ? availableFrames / transitionsCount : 0;
        let currentFrame = 0;
        let lastValue;
        // First value is at time 0
        if (rawKeyframes.length > 0) {
            const first = rawKeyframes[0];
            if (typeof first === 'object' && first && first.type === 'hold') {
                // Should not start with hold? Or hold implies holding previous value (undefined?)
                // Assume valid sequence starts with value
            }
        }
        rawKeyframes.forEach((item, index) => {
            if (typeof item === 'object' && item && item.type === 'hold') {
                const h = item;
                // Extend current time by hold duration
                // Add a keyframe at the end of hold with the SAME last value
                if (lastValue !== undefined) {
                    currentFrame += h.frames;
                    inputRange.push(currentFrame / duration);
                    simpleKeyframes.push(lastValue);
                }
            }
            else {
                const val = item;
                if (lastValue !== undefined) {
                    // This is a new value, so we transitioned from lastValue
                    // Add transition duration
                    currentFrame += transitionDuration;
                }
                // Add point
                // If it's the very first item, currentFrame is 0.
                inputRange.push(currentFrame / duration);
                simpleKeyframes.push(val);
                lastValue = val;
            }
        });
        // Ensure we cover 0 to 1
        // If constructed range doesn't start at 0 (should be handled by logic)
        // If doesn't end at 1?
        // Due to precision or logic, ensure last point is at 1 or we might extrapolate
        return interpolate(progress, inputRange, simpleKeyframes, { easing: easingFn });
    }
    const keyframes = value;
    if (keyframes.length === 1)
        return keyframes[0];
    const inputRange = keyframes.map((_, i) => i / (keyframes.length - 1));
    return interpolate(progress, inputRange, keyframes, { easing: easingFn });
}
export function getEasingFunction(easing) {
    if (!easing)
        return undefined;
    if (typeof easing === "function")
        return easing;
    return Easing[easing];
}
export function buildTransformString(transforms, progress, easingFn, duration) {
    const transformParts = [];
    if (transforms.transform !== undefined) {
        const rawVal = transforms.transform;
        const isString = typeof rawVal === 'string' || (Array.isArray(rawVal) && rawVal.length > 0 && typeof rawVal[0] === 'string');
        if (!isString) {
            const matrixVal = interpolateKeyframes(transforms.transform, progress, easingFn, duration);
            transformParts.push(matrixToCSS(matrixVal));
        }
        else {
            if (typeof rawVal === 'string') {
                transformParts.push(rawVal);
            }
        }
    }
    if (transforms.x !== undefined) {
        const xVal = interpolateKeyframes(transforms.x, progress, easingFn, duration);
        transformParts.push(`translateX(${xVal}px)`);
    }
    if (transforms.y !== undefined) {
        const yVal = interpolateKeyframes(transforms.y, progress, easingFn, duration);
        transformParts.push(`translateY(${yVal}px)`);
    }
    if (transforms.z !== undefined) {
        const zVal = interpolateKeyframes(transforms.z, progress, easingFn, duration);
        transformParts.push(`translateZ(${zVal}px)`);
    }
    if (transforms.scale !== undefined) {
        const scaleVal = interpolateKeyframes(transforms.scale, progress, easingFn, duration);
        transformParts.push(`scale(${scaleVal})`);
    }
    if (transforms.scaleX !== undefined) {
        const scaleXVal = interpolateKeyframes(transforms.scaleX, progress, easingFn, duration);
        transformParts.push(`scaleX(${scaleXVal})`);
    }
    if (transforms.scaleY !== undefined) {
        const scaleYVal = interpolateKeyframes(transforms.scaleY, progress, easingFn, duration);
        transformParts.push(`scaleY(${scaleYVal})`);
    }
    if (transforms.rotate !== undefined) {
        const rotateVal = interpolateKeyframes(transforms.rotate, progress, easingFn, duration);
        transformParts.push(`rotate(${rotateVal}deg)`);
    }
    if (transforms.rotateX !== undefined) {
        const rotateXVal = interpolateKeyframes(transforms.rotateX, progress, easingFn, duration);
        transformParts.push(`rotateX(${rotateXVal}deg)`);
    }
    if (transforms.rotateY !== undefined) {
        const rotateYVal = interpolateKeyframes(transforms.rotateY, progress, easingFn, duration);
        transformParts.push(`rotateY(${rotateYVal}deg)`);
    }
    if (transforms.rotateZ !== undefined) {
        const rotateZVal = interpolateKeyframes(transforms.rotateZ, progress, easingFn, duration);
        transformParts.push(`rotateZ(${rotateZVal}deg)`);
    }
    if (transforms.skew !== undefined) {
        const skewVal = interpolateKeyframes(transforms.skew, progress, easingFn, duration);
        transformParts.push(`skew(${skewVal}deg)`);
    }
    if (transforms.skewX !== undefined) {
        const skewXVal = interpolateKeyframes(transforms.skewX, progress, easingFn, duration);
        transformParts.push(`skewX(${skewXVal}deg)`);
    }
    if (transforms.skewY !== undefined) {
        const skewYVal = interpolateKeyframes(transforms.skewY, progress, easingFn, duration);
        transformParts.push(`skewY(${skewYVal}deg)`);
    }
    return transformParts.join(" ");
}
export function buildMotionStyles(config) {
    const { progress, transforms = {}, styles = {}, easing, baseStyle = {}, duration } = config;
    const easingFn = getEasingFunction(easing);
    const result = { ...baseStyle };
    const transformString = buildTransformString(transforms, progress, easingFn, duration);
    if (transformString) {
        result.transform = transformString;
    }
    if (styles.opacity !== undefined) {
        result.opacity = interpolateKeyframes(styles.opacity, progress, easingFn, duration);
    }
    if (styles.color) {
        // Ensure array
        const colorVal = Array.isArray(styles.color) ? styles.color : [styles.color];
        const safeColors = colorVal.filter(c => typeof c === 'string');
        result.color = interpolateColorKeyframes(safeColors, progress, easingFn);
    }
    if (styles.backgroundColor) {
        const bgVal = Array.isArray(styles.backgroundColor) ? styles.backgroundColor : [styles.backgroundColor];
        const safeColors = bgVal.filter(c => typeof c === 'string');
        result.backgroundColor = interpolateColorKeyframes(safeColors, progress, easingFn);
    }
    if (styles.blur !== undefined) {
        const blurVal = interpolateKeyframes(styles.blur, progress, easingFn, duration);
        if (Number.isFinite(blurVal) && blurVal > 0) {
            result.filter = `blur(${blurVal}px)`;
        }
    }
    if (styles.borderRadius !== undefined) {
        const radiusVal = interpolateKeyframes(styles.borderRadius, progress, easingFn, duration);
        result.borderRadius = radiusVal; // AnimatedValue can be strictly number or number[], but CSS needs 'px' or % often.
        // However interpolateKeyframes returns T. If T is number, it's just number.
        // borderRadius in React can be number (px) or string.
        // Let's assume input is number (px) or string if needed, but interpolateKeyframes handles numbers best.
        // If it returns number, React assumes px.
        // If user provided string keyframes, interpolateKeyframes might fail if it expects numbers, but it handles arrays?
        // interpolateKeyframes casts to T. If T is number, fine.
        // Let's trust React to handle number as px.
    }
    return result;
}
export function useMotionTiming(config) {
    const frame = useCurrentFrame();
    const { fps } = useVideoConfig();
    const stepTiming = useStepTiming();
    const { frames, duration, delay = 0, stagger = 0, unitIndex = 0, cycleOffset, } = config;
    let startFrame;
    let endFrame;
    if (cycleOffset !== undefined) {
        startFrame = 0;
        endFrame = duration ?? 30;
    }
    else if (frames) {
        startFrame = frames[0];
        endFrame = frames[1];
    }
    else if (stepTiming?.stepConfig) {
        const computedDelay = delay + (unitIndex * stagger);
        startFrame = stepTiming.stepConfig.enterFrame + computedDelay;
        const computedDuration = duration ?? 30;
        endFrame = startFrame + computedDuration;
    }
    else {
        startFrame = 0;
        endFrame = duration ?? fps;
    }
    const totalDuration = endFrame - startFrame;
    const baseFrame = cycleOffset !== undefined ? cycleOffset : Math.max(0, frame - delay);
    const relativeFrame = baseFrame - (unitIndex * stagger);
    const progress = Math.min(Math.max((relativeFrame - startFrame) / totalDuration, 0), 1);
    return progress;
}
export { useStepTiming, StepTimingContext } from "../StepContext.js";
