// remotion-bits/dist/utils/interpolate.js
import { Matrix4 } from 'three';
import { interpolateMatrix4 } from "./interpolate3d.js";
import { Transform3D } from "./transform3d.js";
const springFactory = (config = {}) => {
    const { mass = 1, stiffness = 100, damping = 10 } = config;
    const w0 = Math.sqrt(stiffness / mass);
    const zeta = damping / (2 * Math.sqrt(stiffness * mass));
    return (t) => {
        if (t === 0)
            return 0;
        if (t === 1)
            return 1;
        if (zeta < 1) {
            const wd = w0 * Math.sqrt(1 - zeta * zeta);
            return 1 - Math.exp(-zeta * w0 * t) * (Math.cos(wd * t) + (zeta * w0 / wd) * Math.sin(wd * t));
        }
        else {
            return 1 - Math.exp(-w0 * t) * (1 + w0 * t);
        }
    };
};
/**
 * Easing functions for non-linear interpolation
 */
export const Easing = {
    linear: (t) => t,
    easeIn: (t) => t * t,
    easeOut: (t) => t * (2 - t),
    easeInOut: (t) => (t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t),
    easeInQuad: (t) => t * t,
    easeOutQuad: (t) => t * (2 - t),
    easeInOutQuad: (t) => (t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t),
    easeInCubic: (t) => t * t * t,
    easeOutCubic: (t) => (--t) * t * t + 1,
    easeInOutCubic: (t) => t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1,
    easeInSine: (t) => 1 - Math.cos((t * Math.PI) / 2),
    easeOutSine: (t) => Math.sin((t * Math.PI) / 2),
    easeInOutSine: (t) => -(Math.cos(Math.PI * t) - 1) / 2,
    easeInQuart: (t) => t * t * t * t,
    easeOutQuart: (t) => 1 - (--t) * t * t * t,
    easeInOutQuart: (t) => t < 0.5 ? 8 * t * t * t * t : 1 - 8 * (--t) * t * t * t,
    spring: springFactory(),
};
export const steps = (steps) => (t) => Math.floor(t * steps) / steps;
export const spring = springFactory;
export const hold = (frames) => ({ type: 'hold', frames });
export function interpolate(input, inputRange, outputRange, options) {
    const { extrapolateLeft = 'extend', extrapolateRight = 'extend', easing } = options || {};
    const normalizedOutputRange = outputRange.map((value) => value instanceof Transform3D ? value.toMatrix4() : value);
    if (inputRange.length !== normalizedOutputRange.length) {
        throw new Error('inputRange and outputRange must have the same length');
    }
    if (inputRange.length < 2) {
        throw new Error('inputRange must have at least 2 elements');
    }
    // Find the segment where input falls
    let segmentIndex = -1;
    for (let i = 0; i < inputRange.length - 1; i++) {
        const start = inputRange[i];
        const end = inputRange[i + 1];
        if (start === end) {
            if (input === start) {
                segmentIndex = i;
                break;
            }
        }
        else if (start < end) {
            if (input >= start && input <= end) {
                segmentIndex = i;
                break;
            }
        }
        else {
            if (input <= start && input >= end) {
                segmentIndex = i;
                break;
            }
        }
    }
    // Handle extrapolation
    if (segmentIndex === -1) {
        if (input < inputRange[0]) {
            // Extrapolate left
            if (extrapolateLeft === 'clamp') {
                return normalizedOutputRange[0] instanceof Matrix4
                    ? normalizedOutputRange[0].clone()
                    : normalizedOutputRange[0];
            }
            else if (extrapolateLeft === 'identity') {
                return input;
            }
            else {
                // extend
                segmentIndex = 0;
            }
        }
        else {
            // Extrapolate right (input > inputRange[last])
            if (extrapolateRight === 'clamp') {
                const last = normalizedOutputRange[normalizedOutputRange.length - 1];
                return last instanceof Matrix4 ? last.clone() : last;
            }
            else if (extrapolateRight === 'identity') {
                return input;
            }
            else {
                // extend
                segmentIndex = inputRange.length - 2;
            }
        }
    }
    // Interpolate within the segment
    const inputStart = inputRange[segmentIndex];
    const inputEnd = inputRange[segmentIndex + 1];
    const outputStart = normalizedOutputRange[segmentIndex];
    const outputEnd = normalizedOutputRange[segmentIndex + 1];
    // Handle zero-length segments (hold frames)
    if (inputStart === inputEnd) {
        return outputStart instanceof Matrix4 ? outputStart.clone() : outputStart;
    }
    // Calculate progress (0 to 1)
    let progress = (input - inputStart) / (inputEnd - inputStart);
    // Apply easing if provided
    if (easing) {
        const easingFn = typeof easing === 'string' ? Easing[easing] : easing;
        progress = easingFn(progress);
    }
    if (typeof outputStart === 'number' && typeof outputEnd === 'number') {
        return outputStart + (outputEnd - outputStart) * progress;
    }
    if (outputStart instanceof Matrix4 && outputEnd instanceof Matrix4) {
        return interpolateMatrix4(outputStart, outputEnd, progress);
    }
    return outputStart instanceof Matrix4 ? outputStart.clone() : outputStart;
}
interpolate.hold = hold;
/**
 * Resolves an InterpolateValue.
 */
export function resolveInterpolateValue(value, frame) {
    if (Array.isArray(value)) {
        const [inputRange, outputRange, options] = value;
        return interpolate(frame, inputRange, outputRange, options);
    }
    return value;
}
