// remotion-bits/dist/catalog/runtime.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { findBits, getBitCatalogSummaries, getBitCatalogTags, getDocsBitCatalogData, listBitCatalog, resolveBitCatalogIdentifier, } from "./shared.js";
const packageRoot = fileURLToPath(new URL('../../', import.meta.url));
const assertSafeSourcePath = (sourcePath) => {
    const absolutePath = path.resolve(packageRoot, sourcePath);
    const relativePath = path.relative(packageRoot, absolutePath);
    if (relativePath.startsWith('..') || path.isAbsolute(relativePath)) {
        throw new Error(`Bit source path resolves outside the package: ${sourcePath}`);
    }
    return absolutePath;
};
const readBitSource = (sourcePath) => fs.readFileSync(assertSafeSourcePath(sourcePath), 'utf8');
export const fetchBit = async (identifier) => {
    const resolution = resolveBitCatalogIdentifier(identifier);
    if (!resolution.entry) {
        return null;
    }
    return {
        ...resolution.entry,
        sourceCode: readBitSource(resolution.entry.sourcePath),
    };
};
export { findBits, getBitCatalogSummaries, getBitCatalogTags, getDocsBitCatalogData, listBitCatalog, resolveBitCatalogIdentifier, };
