// remotion-bits/dist/components/AnimatedText.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { useCurrentFrame, random, useVideoConfig } from "remotion";
import { useMotionTiming, useStepTiming, buildMotionStyles, getEasingFunction, interpolateKeyframes, } from "../utils/motion/index.js";
// ============================================================================
// HELPER FUNCTIONS
// ============================================================================
function splitText(text, mode) {
    if (mode === "none")
        return [text];
    if (mode === "word")
        return text.split(/(\s+)/);
    if (mode === "character")
        return text.split("");
    if (mode === "line")
        return text.split("\n");
    // Custom separator: split by the provided string
    return text.split(mode);
}
// ============================================================================
// COMPONENT
// ============================================================================
export const AnimatedText = ({ transition, children, className, style, }) => {
    const frame = useCurrentFrame();
    const { x, y, z, scale, scaleX, scaleY, rotate, rotateX, rotateY, rotateZ, skew, skewX, skewY, opacity, color, backgroundColor, blur, frames, duration, delay = 0, split = "none", splitStagger = 0, easing, cycle, } = transition ?? {};
    const { fps } = useVideoConfig();
    const stepTiming = useStepTiming();
    const totalDuration = React.useMemo(() => {
        if (frames)
            return frames[1] - frames[0];
        if (stepTiming?.stepConfig)
            return duration ?? 30;
        return duration ?? fps;
    }, [frames, stepTiming, duration, fps]);
    const easingFn = getEasingFunction(easing);
    // Handle text cycling
    let content;
    let cycleFrameOffset;
    if (cycle) {
        const { texts, itemDuration } = cycle;
        const relativeFrame = Math.max(0, frame - delay);
        const cycleIndex = Math.floor(relativeFrame / itemDuration) % texts.length;
        content = texts[cycleIndex] || "";
        // Calculate frame offset within the current cycle item
        cycleFrameOffset = relativeFrame % itemDuration;
    }
    else {
        content = typeof children === "string" ? children : String(children || "");
    }
    // Split text
    const units = splitText(content, split);
    // Render function for each unit
    const renderUnit = (unit, index) => {
        // Calculate progress using motion framework
        const progress = useMotionTiming({
            frames,
            duration,
            delay,
            stagger: splitStagger,
            unitIndex: index,
            easing,
            cycleOffset: cycleFrameOffset,
        });
        // Build animated styles using motion framework
        const unitStyle = buildMotionStyles({
            progress,
            transforms: transition,
            styles: { opacity, color, backgroundColor, blur },
            easing: easingFn,
            baseStyle: {
                display: "inline-block",
                whiteSpace: "pre",
            },
            duration: totalDuration,
        });
        // Handle glitch effect
        const glitchVal = transition?.glitch !== undefined
            ? interpolateKeyframes(transition.glitch, progress, easingFn, totalDuration)
            : 0;
        let displayUnit = unit;
        if (glitchVal > 0) {
            const GLITCH_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*()_+-=[]{}|;:,.<>?/~`";
            displayUnit = unit.split("").map((char, charIdx) => {
                if (char === " " || char === "\n")
                    return char;
                const seed = `glitch-${frame}-${index}-${charIdx}`;
                if (random(seed) < glitchVal) {
                    const charSeed = `char-${frame}-${index}-${charIdx}`;
                    const i = Math.floor(random(charSeed) * GLITCH_CHARS.length);
                    return GLITCH_CHARS[i];
                }
                return char;
            }).join("");
        }
        return (_jsx("span", { style: unitStyle, children: displayUnit }, index));
    };
    return (_jsx("span", { className: className, style: {
            display: "inline-block",
            ...style,
        }, children: units.map(renderUnit) }));
};
