// remotion-bits/dist/components/AnimatedCounter.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from "react";
import { useVideoConfig } from "remotion";
import { useMotionTiming, useStepTiming, buildMotionStyles, interpolateKeyframes, getEasingFunction, } from "../utils/motion/index.js";
// ============================================================================
// COMPONENT
// ============================================================================
export const AnimatedCounter = ({ transition, prefix, postfix, toFixed = 0, className, style, }) => {
    const { fps } = useVideoConfig();
    const stepTiming = useStepTiming();
    const { values, frames, duration, delay = 0, easing, opacity, color, backgroundColor, blur, ...transforms } = transition;
    // Calculate total duration for interpolation context
    const totalDuration = React.useMemo(() => {
        if (frames)
            return frames[1] - frames[0];
        if (stepTiming?.stepConfig)
            return duration ?? 30;
        return duration ?? fps;
    }, [frames, stepTiming, duration, fps]);
    const easingFn = getEasingFunction(easing);
    // Calculate progress
    const progress = useMotionTiming({
        frames,
        duration,
        delay,
        easing: easingFn,
    });
    // Interpolate the counter value
    const value = interpolateKeyframes(values, progress, easingFn, totalDuration);
    // Build container styles
    const containerStyle = buildMotionStyles({
        progress,
        transforms,
        styles: { opacity, color, backgroundColor, blur },
        easing: easingFn,
        baseStyle: {
            display: "inline-flex",
            alignItems: "center",
            ...style
        },
        duration: totalDuration,
    });
    return (_jsxs("span", { className: className, style: containerStyle, children: [prefix, _jsx("span", { children: value.toFixed(toFixed) }), postfix] }));
};
