// remotion-bits/dist/components/StaggeredMotion.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { random } from "remotion";
import { useMotionTiming, buildMotionStyles, getEasingFunction, } from "../utils/motion/index.js";
function calculateStaggerIndex(actualIndex, totalChildren, direction) {
    if (direction === "reverse") {
        return totalChildren - 1 - actualIndex;
    }
    if (direction === "center") {
        const mid = Math.floor(totalChildren / 2);
        return Math.abs(actualIndex - mid);
    }
    if (direction === "random") {
        const indices = Array.from({ length: totalChildren }, (_, i) => i);
        for (let i = indices.length - 1; i > 0; i--) {
            const j = Math.floor(random(`stagger-${i}`) * (i + 1));
            [indices[i], indices[j]] = [indices[j], indices[i]];
        }
        return indices.indexOf(actualIndex);
    }
    return actualIndex;
}
export const StaggeredMotion = ({ transition, children, className, style, cycleOffset, }) => {
    const { x, y, z, scale, scaleX, scaleY, rotate, rotateX, rotateY, rotateZ, skew, skewX, skewY, opacity, color, backgroundColor, blur, borderRadius, frames, duration, delay = 0, stagger = 0, staggerDirection = "forward", easing, } = transition ?? {};
    const easingFn = getEasingFunction(easing);
    const childArray = React.Children.toArray(children);
    const totalChildren = childArray.length;
    const renderChild = (child, actualIndex) => {
        const staggerIndex = calculateStaggerIndex(actualIndex, totalChildren, staggerDirection);
        const resolvedDuration = duration ?? (frames ? frames[1] - frames[0] : 30);
        const progress = useMotionTiming({
            frames,
            duration,
            delay,
            stagger,
            unitIndex: staggerIndex,
            easing,
            cycleOffset: cycleOffset !== undefined ? cycleOffset - delay : undefined,
        });
        const motionStyle = buildMotionStyles({
            progress,
            transforms: transition,
            styles: { opacity, color, backgroundColor, blur, borderRadius },
            easing: easingFn,
            duration: resolvedDuration,
        });
        if (React.isValidElement(child)) {
            const existingStyle = child.props.style || {};
            return React.cloneElement(child, {
                key: actualIndex,
                style: { ...existingStyle, ...motionStyle },
            });
        }
        return (_jsx("span", { style: motionStyle, children: child }, actualIndex));
    };
    return (_jsx("div", { className: className, style: style, children: childArray.map(renderChild) }));
};
